plugins/promoter/server.go

```diff
@@ -0,0 +1,309 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/pluginhelp"
+)
+
+type githubClient interface {
+	CreatePullRequest(org, repo, title, body, head, base string, canModify bool) (int, error)
+	CreateComment(org, repo string, number int, comment string) error
+}
+
+// HelpProvider construct the pluginhelp.PluginHelp for this plugin.
+func HelpProvider(enabledRepos []string) (*pluginhelp.PluginHelp, error) {
+	pluginHelp := &pluginhelp.PluginHelp{
+		Description: `The promoter plugin is used for promoting PRs across branches. For every successful promoter invocation a new PR is opened against the target branch and assigned to the requester.`,
+	}
+	pluginHelp.AddCommand(pluginhelp.Command{
+		Description: "Promote changes to a different branch. This plugin automatically promotes PRs once merged to the configured target branch.",
+	})
+	return pluginHelp, nil
+}
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	tokenGenerator   func() []byte
+	botName          string
+	botPassGenerator func() []byte
+	email            string
+
+	gc *git.Client
+	// Used for unit testing
+	push func(org, repo, dir, branch string) error
+	ghc  githubClient
+	git  string
+	log  *logrus.Entry
+	wg   *sync.WaitGroup
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	eventType, eventGUID, payload, ok, _ := github.ValidateWebhook(w, r, s.tokenGenerator())
+	if !ok {
+		return
+	}
+
+	sources, targets, ok := validateParams(w, r)
+	if !ok {
+		return
+	}
+
+	if err := s.handleEvent(eventType, eventGUID, payload, sources, targets); err != nil {
+		s.log.WithError(err).Error("Error parsing event.")
+	}
+}
+
+func validateParams(w http.ResponseWriter, r *http.Request) ([]string, []string, bool) {
+	params := r.URL.Query()
+
+	sources, ok := params["source"]
+	if !ok {
+		responseHTTPError(w, http.StatusBadRequest, "400 Bad Request: Missing source parameter")
+		return []string{}, []string{}, false
+	}
+
+	targets, ok := params["target"]
+	if !ok {
```

Comment:
Shouldn't we be checking whether the `source` and `target` params are non-empty as well?

Reply:
Added a check